framework/src/interfaces/Coupleable.C

```diff
@@ -825,12 +832,115 @@ Coupleable::coupledArrayValues(const std::string & var_name) const
   return coupledVectorHelper<const ArrayVariableValue *>(var_name, func);
 }
 
+MooseVariable &
+Coupleable::writableVariable(const std::string & var_name, unsigned int comp)
+{
+  auto * var = dynamic_cast<MooseVariable *>(getVar(var_name, comp));
+
+  const auto * aux = dynamic_cast<const AuxKernel *>(this);
+  const auto * euo = dynamic_cast<const ElementUserObject *>(this);
+  const auto * nuo = dynamic_cast<const NodalUserObject *>(this);
+
+  if (!aux && !euo && !nuo)
+    mooseError("writableVariable() can only be called from AuxKernels, ElementUserObjects, or "
+               "NodalUserObjects. '",
+               _obj->name(),
+               "' is neither of those.");
+
+  if (aux && !aux->isNodal() && var->isNodal())
+    mooseError("The elemental AuxKernel '",
+               _obj->name(),
+               "' cannot obtain a writable reference to the nodal variable '",
+               var->name(),
+               "'.");
+  if (euo && var->isNodal())
+    mooseError("The ElementUserObject '",
+               _obj->name(),
+               "' cannot obtain a writable reference to the nodal variable '",
+               var->name(),
+               "'.");
+
+  // make sure only one object can access a variable
+  checkWritableVar(var);
+
+  return *var;
+}
+
 VariableValue &
 Coupleable::writableCoupledValue(const std::string & var_name, unsigned int comp)
 {
+  mooseDeprecated("Coupleable::writableCoupledValue is deprecated, please use "
+                  "Coupleable::writableVariable instead. ");
+
+  // check if the variable exists
+  auto * const var = getVar(var_name, comp);
+  if (!var)
+    mooseError(
+        "Unable to create a writable reference for '", var_name, "', is it a constant expression?");
+
+  // is the requested variable an AuxiliaryVariable?
+  if (!_c_fe_problem.getAuxiliarySystem().hasVariable(var->name()))
+    mooseError(
+        "'", var->name(), "' must be an auxiliary variable in Coupleable::writableCoupledValue");
+
+  // check that the variable type (elemental/nodal) is compatible with the object type
+  const auto * aux = dynamic_cast<const AuxKernel *>(this);
+
+  if (!aux)
+    mooseError("writableCoupledValue() can only be called from AuxKernels, but '",
+               _obj->name(),
+               "' is not an AuxKernel.");
+
+  if (!aux->isNodal() && var->isNodal())
+    mooseError("The elemental AuxKernel '",
+               _obj->name(),
+               "' cannot obtain a writable reference to the nodal variable '",
+               var->name(),
+               "'.");
+
+  // make sure only one object can access a variable
+  checkWritableVar(var);
+
   return const_cast<VariableValue &>(coupledValue(var_name, comp));
 }
 
+void
+Coupleable::checkWritableVar(MooseVariable * var)
+{
+  // check block restrictions for compatibility
```

Comment:
It's okay if `var` has more blocks than `br` right?

Reply:
yeah, that's ... compatible